### PR TITLE
bpo-38208: Simplify string.Template by using __init_subclass__().

### PR DESCRIPTION
**Pull request by serhiy-storchaka**
https://bugs.python.org/issue38208

### CODE REVIEW

File: Lib/dataclasses.py [hunk 1]
```diff
@@ -212,6 +208,9 @@ def __init__(self, type):
     def __repr__(self):
         return f'dataclasses.InitVar[{self.type.__name__}]'
 
+    def __class_getitem__(cls, params):
```
Comment (warsaw) [inline Lib/dataclasses.py]: These changes aren't related to the title of this PR.   Please either split it out into a separate PR, or retitle this PR.

File: Lib/string.py [hunk 2]
```diff
@@ -89,6 +66,25 @@ class Template(metaclass=_TemplateMetaclass):
     braceidpattern = None
     flags = _re.IGNORECASE
 
+    def __init_subclass__(cls):
```
Comment (warsaw) [inline Lib/string.py:69]: This seems like a better approach, but if you're modernizing this then I have a few other suggestions.

File: Lib/string.py [hunk 3]
```diff
@@ -89,6 +66,25 @@ class Template(metaclass=_TemplateMetaclass):
...
+        super().__init_subclass__()
+        if 'pattern' in cls.__dict__:
+            pattern = cls.pattern
+        else:
+            pattern = r"""
```
Comment (warsaw) [inline Lib/string.py]: Perhaps this pattern ought to be put at module scope, just above the class definition?  The verbose representation might be more readable without all that leading whitespace.

File: Lib/string.py [hunk 4]
```diff
@@ -89,6 +66,25 @@ class Template(metaclass=_TemplateMetaclass):
...
+            %(delim)s(?:
+              (?P<escaped>%(delim)s) |   # Escape sequence of two delimiters
+              (?P<named>%(id)s)      |   # delimiter and a Python identifier
+              {(?P<braced>%(bid)s)}  |   # delimiter and a braced identifier
+              (?P<invalid>)              # Other ill-formed delimiter exprs
+            )
+            """ % {
```
Comment (warsaw) [inline Lib/string.py]: Too bad we can't convert this to a `.format()` string without breaking backward compatibility.

File: Lib/string.py [hunk 5]
```diff
@@ -146,6 +142,7 @@ def convert(mo):
                              self.pattern)
         return self.pattern.sub(convert, self.template)
 
+Template.__init_subclass__()  # setup pattern
```
Comment (warsaw) [inline Lib/string.py]: This is a bit of a head scratcher, but I think the comment could be improved to make it more clear.  What's really going on is that this is required to initialize `pattern` when no subclass is created.  Can you improve the clarity of the comment (and maybe move it above this line, and make it a full sentence)?   If you need a suggestion, let me know.

**Review by warsaw — CHANGES_REQUESTED** [review]

File: Lib/dataclasses.py (hunk 1 above)
Comment (serhiy-storchaka) [reply Lib/dataclasses.py]: Oh, yes. They are from other issue.

File: Lib/string.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/string.py]: We can. We can even convert it to f-string. I tried, but returned to the old format.

File: Lib/string.py (hunk 5 above)
Comment (serhiy-storchaka) [reply Lib/string.py]: I would appreciate your help with a comment.

File: Lib/string.py (hunk 4 above)
Comment (serhiy-storchaka) [reply Lib/string.py]: The only problem with `.format()` and f-string is that we nee to duplicate literal braces. And the template/f-string will look strange, because braces have special meaning in regexpes and in new format.

Comment (serhiy-storchaka) [reply Lib/string.py]: With f-string it would look:
```python
    def __init_subclass__(cls):
        super().__init_subclass__()
        if 'pattern' in cls.__dict__:
            pattern = cls.pattern
        else:
            delim = _re.escape(cls.delimiter)
            id = cls.idpattern
            bid = cls.braceidpattern or cls.idpattern
            pattern = fr"""
            {delim}(?:
              (?P<escaped>{delim})  |   # Escape sequence of two delimiters
              (?P<named>{id})       |   # delimiter and a Python identifier
              {{(?P<braced>{bid})}} |   # delimiter and a braced identifier
              (?P<invalid>)             # Other ill-formed delimiter exprs
            )
            """
        cls.pattern = _re.compile(pattern, cls.flags | _re.VERBOSE)
```

**Comment by serhiy-storchaka** [comment]
@warsaw, what is your thought about using an f-string?

**Review by Budda0ne — COMMENTED** [review]
Ok

**Review by Budda0ne — COMMENTED** [review]
Ok
